YunoHost/yunohost
# [enh] Improve service status checks (and service add CLI)

## The problem

Many small issues in the current system that test the service status : 
- postfix's status check should also look at `postfix@-` (which is a different service from `postfix`)
- `yunohost-firewall`'s status is `exited` because it's just a script that generates iptables rules ...
- `active` vs. `status` is a mess ... what do they mean ? why can it be `active` while being `exited` or viceversa ?
- `loaded` is a shit name (even though that's the one used by systemd). What it means is if it should start on boot so let's call it that way
- `activated_at` is only meaningful if the service is running ...
- we should have a general system to test services configuration (e.g. `nginx -t` etc...)
- we have no relation between ports and services, which could be useful for diagnosis
- many app add a service without providing a description, whereas we should encourage packagers to provide one
- `service log` just returns `/var/log/daemon.log` and `/var/log/syslog` for many services, which is useless because those files are too general instead of being specific to a single service.

## Solution

- Have an optional `test-status` in `services.yml` to define how to test the status of a service. For example, `systemctl show postfix@- | grep -q "^SubState=running"` for postfix.
- Remove the `active` state ... just `status` (expected to be `running`) is good enough now that we have a proper `test-status` for postfix and yunohost-firewall
- Have an optional `test-conf` in `services.yml` to define what command to use to check if a service conf is valid or broken. So far implemented for nginx, php-fpm, ssh. Didn't check every services capabilities, can check this later...
- `loaded` -> `start_on_boot`
- `activated_at` -> `last_state_change` (using the corresponding systemd info)
- add some `needs_ports_exposed` option for services to keep track of which services needs port exposure
- display a warning if a service is added without providing a description
- propagate all this to the new diagnosis system
- add options `--test_status`, `--test_conf` and `--needs_exposed_ports` in the `yunohost service add`  command line
- make `service log` returns the `journalctl` output by default.

## PR Status

Tested and working on my side ... needs to be propagated on the Services pages in the webadmin

## How to test

Uh make sure to regen-conf yunohost, then you can run `yunohost service status` and `yunohost diagnosis services`

Try to break nginx's conf and see what happens.

## Validation
